docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,833 @@
-
-There are several prerequisites for deleting a channel, we will first consider whether the channel is indexed, index, or rate-based.
-
-If a channel is rate-based or indexed, deleting it will have no effect on other data in the Cesium database, so we can proceed freely.
-
-If a channel is an index channel, then deleting it will potentially cause other channels. Therefore, we must check if any other channels use it for indexing – if any do, then we cannot delete it. A possible race condition here is that while no current channel uses the to-be-deleted channel as an index, we may be in the process of making a new channel that does. This race condition is avoided by `db.mu.Lock()` in both `DeleteChannel()` and `openVirtualOrUnary()`.
-
-Next, we will consider whether there is an entity trying to modify the channel, i.e. a writer or iterator. We will touch on the case of streamers later. Overall, we may only delete a unary or virtual database when the number of open iterators or writers on it is 0.
-
-Firstly, writers: when we open a writer, internally, a call to `newStreamWriter()` followed by `wrapStreamWriter()` is made. More specifically, in `newStreamWriter()`, we call the unary or virtual `openWriter` method, which increments 1 to the `openIteratorWriters` atomic count. We decrement 1 from this count when we close a writer associated with a `unaryDB`. Possible race conditions here are guarded by the mutex lock throughout the entire `newStreamWriter` method.
-
-Iterators: they follow a similar logic: internally a call to `newStreamIterator` is made, and the `openIteratorWriters` count is increased during `uDB.OpenIterator` and decreased when that iterator is closed.
-
-Lastly, we will discuss streamers: fundamentally, streamers do not make modifications to the database and can either be used to stream data being written to it or read from it. We can see that it is okay to have a streamer open on an inexistent unary database, as there will be no write or read requests made to it, so the streamer will receive nothing in its inlet. Eventually, this inlet will be closed.
-
-To recap, we can delete a channel if there are no open iterators or writers on it, and if it is not an index channel for other channels in the database.
+Concurrency is a double-edged sword: while it could make Cesium blazing-fast, it introduces
+additional complexity through parallelism – when many subprocesses run at the same time,
+they create conflicts, or race conditions, when they attempt to access common resources.
+As Cesium looks to become production-ready, it is crucial that Cesium must support
+Read, Write, and Delete (through a tombstone & garbage collection system) operations
+at high speeds without them conflicting each other. One can imagine the possible
+consequences for unhandled race conditions in extreme scenarios such as launch control systems:
+it would not be desirable.
+
+# 4- Analysis
```

Comment:
I went with this numbering system because in all the other RFCs 1, 2, 3 are always vocabulary, motivation, analysis, and design, with a very heavy emphasis on the 3rd and 4th sections. I didn't know if it made sense for this RFC to separate by topic instead of following the pattern we have, what do you think?